packages/codecs-strings/src/base16.ts

```diff
@@ -6,18 +6,63 @@ import {
     VariableSizeDecoder,
     VariableSizeEncoder,
 } from '@solana/codecs-core';
+import { SOLANA_ERROR__CODECS__INVALID_STRING_FOR_BASE, SolanaError } from '@solana/errors';
 
-import { assertValidBaseString } from './assertions';
+const enum HexC {
+    ZERO = 48, // 0
+    NINE = 57, // 9
+    A_UP = 65, // A
+    F_UP = 70, // F
+    A_LO = 97, // a
+    F_LO = 102, // f
+}
+
+const INVALID_STRING_ERROR_BASE_CONFIG = {
+    alphabet: '0123456789abcdef',
+    base: 16,
+} as const;
+
+function charCodeToBase16(char: number) {
+    if (char >= HexC.ZERO && char <= HexC.NINE) return char - HexC.ZERO;
+    if (char >= HexC.A_UP && char <= HexC.F_UP) return char - (HexC.A_UP - 10);
+    if (char >= HexC.A_LO && char <= HexC.F_LO) return char - (HexC.A_LO - 10);
+}
 
 /** Encodes strings in base16. */
 export const getBase16Encoder = (): VariableSizeEncoder<string> =>
     createEncoder({
         getSizeFromValue: (value: string) => Math.ceil(value.length / 2),
         write(value: string, bytes, offset) {
-            const lowercaseValue = value.toLowerCase();
-            assertValidBaseString('0123456789abcdef', lowercaseValue, value);
```

Comment:
I was really surprised that this up-front check was slower than checking for bad characters inline, but it's a _lot_ slower cc/ @lorisleiva.

Reply:
I didn't investigate to see if it was the `toLowercase()` or the string walk that was the slow part.

Reply:
Just caught up on this PR. That's really nice! I love it when we can improve performance without any API trade-off.